### DISCUSSION TITLE
How to enable company-mode in some mode without auto completion?

### DISCUSSION THREAD

**Discussion by haoisli9** [company-mode/company-mode#1323] [Q&A]
how can i enable company mode in some mode without auto completion? i mean, company mode enabled in eshell buffer but do not popup completion frame automaticly. i can toggle completion manually by typing tab etc.

**Comment by yugaego** [comment]
The following snippet works for me, disabling pop-up for `eshell-mode` only:

```elisp
(require 'company)

(global-company-mode 1)
(setq company-minimum-prefix-length 1)

(defun my-eshell-mode-company ()
  (setq-local company-tooltip-idle-delay 100
              company-frontends
              '(company-pseudo-tooltip-unless-just-one-frontend-with-delay
                company-preview-frontend
                company-echo-metadata-frontend)))

(add-hook 'eshell-mode-hook 'my-eshell-mode-company)

(defun my-company-keymap--filter-by-eshell (command)
  (if (eq major-mode 'eshell-mode)
      command
    'company-complete-common))

(define-key company-active-map (kbd "<tab>") ; or TAB
            '(menu-item ""
                        company-complete-common-or-show-delayed-tooltip
                        :filter my-company-keymap--filter-by-eshell))
```

Optionally, instead of filtering, bind a key not used by `company-active-map`:
```elisp
(define-key company-active-map (kbd "M-/")
  #'company-complete-common-or-show-delayed-tooltip)
```
 
Additional information can be found at:
 -  [Wiki Tips & Tricks](https://github.com/company-mode/company-mode/wiki/Tips-&-tricks#configure-different-backends-for-different-modes). 
 - [Disable auto-popup discussion](https://github.com/company-mode/company-mode/discussions/1214).
 - [Emacs mode-specific custom key bindings](https://stackoverflow.com/questions/9818307/emacs-mode-specific-custom-key-bindings-local-set-key-vs-define-key).

**Comment by dgutov** [comment]
@yugaego presented an extensive opinionated approach, but here's also a simpler answer without extra tweaks:

```el
(require 'company)

;; Disable automatic completion (*the answer*)

(setq company-idle-delay nil)

;; Set up tab binding to complete

(define-key company-mode-map [tab] #'company-complete-common)
(define-key company-mode-map (kbd "TAB") #'company-complete-common)

;; Only enable in eshell-mode

(add-hook 'eshell-mode-hook #'company-mode)

;; Or you could enable it globally

;; (global-company-mode 1)
```

**Reply by yugaego** [reply]
Yes, this is simpler approach. I had a bit different objective: keep automatically raised `company` pop-up, getting rid of it only in a specific mode.  Now added a note about it to that answer.
